compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,284 @@
+        # Only the first for-loop is considered, the algorithm can be repeated to elevate the alloc-op higher.
+        if can_move_operation(main_op):
+            for_op = find_parent_for_loop(main_op)
+            main_op.detach()
+
+            rewriter.insert_op(main_op, InsertPoint.before(for_op))
+
+
+class MoveMemrefDims(RewritePattern):
+    """
+    This class represents a rewrite pattern for moving Dim operations outside
+    a (nested) for-loop. This is possible if the dimensions of a memref are not dependent
+    on the loop variables, or a maximum possible value can be derived.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, dim_op: memref.Dim, rewriter: PatternRewriter):
+        def before_loop(op: Operation) -> bool:
+            """
+            Check if the operation is defined before the loop.
+            """
+            return find_parent_for_loop(op) is not find_parent_for_loop(dim_op)
+
+        def can_be_constant(val: SSAValue) -> bool:
+            """
+            Check if all operands are defined outside for loop or can be derived.
+            """
+
+            if isinstance(val, int):
+                return True
+            if isinstance(val, AffineConstantExpr):
+                return True
+
+            # Use the owner of the value to check what kind of operation it is
+            expr = val.owner
+            if before_loop(expr):
+                return True
+            return False
+
+        def dimension_outside_loop(dim_op: memref.Dim) -> bool:
+            """
+            Check if the dimension can be derived outside of this for-loop.
+            """
+            if before_loop(dim_op):
+                return True
+            memref_op = dim_op.source.owner
+            index = int(dim_op.index.owner.value.value.data)
+            return memref_op_outside_loop(memref_op, index)
+
+        def memref_op_outside_loop(memref_op: Operation, index: int) -> bool:
+            if isinstance(memref_op, Block):
+                # This happens when the dim is called on an input argument
+                return True
+            if isinstance(memref_op, memref.Subview):
+                new_op = memref_op.sizes[index].owner
+                if isinstance(new_op, arith.Constant) or isinstance(
+                    new_op, affine.MinOp
+                ):
+                    return True
+                elif isinstance(new_op, memref.Dim):
+                    return dimension_outside_loop(new_op)
+                else:
+                    return False
+            if isinstance(memref_op, linalg.MatmulOp):
+                new_memref_op = memref_op.inputs[index].owner
+                return memref_op_outside_loop(new_memref_op, index)
+            # TODO: Add support for other memref operations, like matmul
+            return False
+
+        def used_by_not_alloc_subview(dim_op: memref.Dim) -> bool:
```

Comment:
This reads a bit easier IMHO
```suggestion
        def used_by_neither_alloc_nor_subview(dim_op: memref.Dim) -> bool:
```